torchvision/prototype/datasets/_builtin/sbd.py

```diff
@@ -103,8 +103,11 @@ def _datapipe(self, resource_dps: List[IterDataPipe]) -> IterDataPipe[Dict[str,
             buffer_size=INFINITE_BUFFER_SIZE,
             drop_none=True,
         )
-        if self._split == "train_noval":
-            split_dp = extra_split_dp
+        split_dp, to_be_closed_dp = (
+            (extra_split_dp, split_dp) if self._split == "train_noval" else (split_dp, extra_split_dp)
+        )
+        for _, file in to_be_closed_dp:
+            file.close()
```

Comment:
This is somewhat problematic in the `split == "train_noval"` case. We want to close all file handles that are coming from the `split_dp`. Unfortunately, `split_dp` comes from a `Demultiplexer`. Thus, by fully iterating over it here, we are loading everything into the demux buffer.

Is there an idiom to "mark" a datapipe to be closed at runtime even if we don't return the datapipe? The only thing I came up with is changing the classifier function of the `Demultiplexer` to drop the samples that would go into `split_dp` if `split == "train_noval"`.

Reply:
It is incorrect to consume Datapipe in construction time. 

```python

        for _, file in to_be_closed_dp:
            file.close()

```

It will not affect executed graph.

Ideally, we want to have something like dp.close(), which will effectively remove dangling pieces of the graph.

But for now you can either use trick like `split_dp = split_dp.concatinate(to_be_closed_dp.filter(lambda x: False))` or do code branching before Demux and avoid creating dangling pieces.

Reply:
> It is incorrect to consume Datapipe in construction time.

I don't think that is fully correct for our case. At construction, we load a couple of files and weave them together into one datapipe. These files are always loaded unconditionally, but for some configurations not all of the files are needed. So we should be able to simply close them during the construction of the dataset datapipe, since they will never make it in the graph, correct?

I agree, we shouldn't do this for datapipes that stem for a `Demultiplexer` if the other parts make it into the final graph.

Reply:
> do code branching before Demux and avoid creating dangling pieces.

If possible, I think that is the better solution, since the other still iterates over all items when the loop should actually be done. I guess that could be irritating as well.

I implemented branching in afb0ec2e718b74d96f4641812e574601b9fbd39b. PTAL